
Some job-control hacking imported from gentoo. Fixes infinte waiting for
executed jobs.

  - Rene Rebe <rene@exaxctcode.de>

--- bash-3.0/tests/jobs4.sub.jobs	2004-05-06 14:02:41.000000000 +0100
+++ bash-3.0/tests/jobs4.sub	2004-09-08 16:07:25.305385245 +0100
@@ -18,5 +18,5 @@
 
 wait
 
-cat &
+sleep 100 &
 kill -1 %% && echo i killed it || echo could not kill it
--- bash-3.0/jobs.c.jobs	2004-09-08 16:07:25.081427747 +0100
+++ bash-3.0/jobs.c	2004-09-08 16:18:06.698604033 +0100
@@ -2475,6 +2475,7 @@
   PROCESS *child;
   pid_t pid;
   int call_set_current, last_stopped_job, job, children_exited, waitpid_flags;
+  static int wcontinued_not_supported = 0;
 
   call_set_current = children_exited = 0;
   last_stopped_job = NO_JOB;
@@ -2488,7 +2489,15 @@
 			: 0;
       if (sigchld || block == 0)
 	waitpid_flags |= WNOHANG;
+    retry:
+      if (wcontinued_not_supported)
+	waitpid_flags &= ~WCONTINUED;
       pid = WAITPID (-1, &status, waitpid_flags);
+      if (pid == -1 && errno == EINVAL)
+	{
+	  wcontinued_not_supported = 1;
+	  goto retry;
+	}
 
       /* The check for WNOHANG is to make sure we decrement sigchld only
 	 if it was non-zero before we called waitpid. */
